secretariat/management/commands/import-from-outline.py

```diff
@@ -0,0 +1,81 @@
+from django.contrib.auth import get_user_model
+from django.core.management.base import BaseCommand
+from django.db.utils import IntegrityError
+
+from config.settings import OUTLINE_URL
+from secretariat.utils.outline import Client as OutlineClient
+from secretariat.utils.outline import RemoteServerError
+
+User = get_user_model()
+
+
+class Command(BaseCommand):
+    help = "Imports users from Outline. Creates missing users in Django, updates existing users if needed."
+
+    def handle(self, *args, **options):
+        client = OutlineClient()
+
+        self.stdout.write(f"Importing users from instance {OUTLINE_URL} .")
+
+        try:
+            outline_users = client.list_users()
+        except RemoteServerError:
+            self.stdout.write(self.style.ERROR("Cannot reach remote server."))
+            exit()
+
+        known_emails = [value[0] for value in User.objects.values_list("email")]
```

Comment:
Si on veut faire de la petite optimisation, on peut utiliser un _set_ plutôt qu'une liste, je dirais qu'on est pile dans un cas d'usage pertinent :
 
```suggestion
        known_emails = set(value[0] for value in User.objects.values_list("email"))
```

Différence entre set et liste : 
- le set contient forcément des éléments distincts,
- le set ne garde pas de notion d'ordre,
- c'est beaucoup plus rapide de chercher dans un set que dans une liste.

C'est de la "petite" optimisation car on n'a qu'une centaine d'e-mails, mais on va quand même chercher dedans à chaque itération de la boucle ligne 29.

Garde en tête que les listes Python ne sont pas spécialement optimisées pour chercher dedans. Ça sera plus utile pour l'advent of code que pour ce script, mais c'est toujours ça. ^^

Reply:
J'ai pas encore ces notions là, grand merci pour l'explication. Je me fais une flashcard sur logseq pour le mémoriser 🤓